### DISCUSSION TITLE
Is there a way to limit the memory usage of a gunicorn worker with FastAPI?

### DISCUSSION THREAD

**Discussion by marcos-mv** [fastapi/fastapi#13009] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
gunicorn.conf.py


import multiprocessing

bind = "0.0.0.0:8000"
workers = 10
worker_class = "uvicorn.workers.UvicornWorker"
timeout = 120
max_requests = 100
max_requests_jitter = 5
proc_name = "intranet"
```

Dockerfile

```
# Dockerfile.prod

# pull the official docker image
FROM python:3.10.8-slim

ARG GITHUB_USERNAME
ARG GITHUB_PERSONAL_ACCESS_TOKEN

# set work directory
WORKDIR /app

RUN mkdir -p /mnt/storage
RUN mkdir /app/logs

# set enviroments
ENV GENERATE_SOURCEMAP=false
ENV TZ="America/Sao_Paulo"

RUN apt-get update \
  && apt-get -y install git \
  && apt-get clean

# install dependencies
COPY requirements.txt .
RUN pip install -r requirements.txt


# copy project
COPY . .

EXPOSE 8000

CMD ["gunicorn", "orquestrador:app", "-k", "worker.MyUvicornWorker"]
```

### Description

This is my gunicorn.conf.py file. I’d like to know if it’s possible to set a memory limit for each worker. I’m running a FastAPI application in a Docker container with a 5 GB memory cap. The application has 10 workers, but I’m experiencing a memory leak issue: one of the workers eventually exceeds the container's memory limit, causing extreme slowdowns until the container is restarted. Is there a way to limit each worker's memory consumption to, for example, 1 GB? Thank you in advance.

I looked at the gunicorn documentation, but I didn't find any mention of a worker's memory limitation.

This video is from inside my fastapi container with htop to see the behavior of the gunicorn worker. 

https://github.com/user-attachments/assets/a4929b06-b975-432c-b42a-6dcc92e38d35

This image shows the moment when a worker from the uvicorn explodes
![memory prblem](https://github.com/user-attachments/assets/783c1d5e-81f8-4882-b1aa-3ac6cd1871d2)

### Operating System

Linux

### FastAPI Version

0.73.0

### Pydantic Version

1.8.2

### Python Version

3.10.8

**Comment by AliYmn** [comment]
Hey there! I understand you're having memory issues with your FastAPI application. Let me offer two simple but effective solutions:

## Solution 1: Docker Memory Limits with Worker Restart

This solution combines Docker's memory limits with Gunicorn's worker restart feature:

1. Update your gunicorn.conf.py:
```python
# gunicorn.conf.py
bind = "0.0.0.0:8000"
workers = 10
worker_class = "uvicorn.workers.UvicornWorker"
max_requests = 1000      # Restart worker after handling 1000 requests
max_requests_jitter = 50 # Add small randomization to prevent simultaneous restarts
```

2. Run Docker with memory limits:
```bash
docker run -d \
  --memory="5g" \
  --memory-swap="5g" \
  --memory-reservation="4.5g" \
  your-image-name
```

This approach:
- Automatically restarts workers after handling 1000 requests, preventing memory buildup
- Sets hard memory limits at the container level
- Is easy to implement and maintain

## Solution 2: Process-Level Memory Management

This solution focuses on monitoring and managing memory at the process level:

1. Create a custom worker (save as worker.py):
```python
from uvicorn.workers import UvicornWorker
import psutil
import os

class MemoryLimitedWorker(UvicornWorker):
    def check_memory(self):
        process = psutil.Process(os.getpid())
        memory_usage = process.memory_info().rss / 1024 / 1024  # MB
        if memory_usage > 450:  # 450MB limit per worker
            self.alive = False
            self.log.warning(f"Worker exceeded memory limit ({memory_usage}MB). Restarting...")
    
    async def handle(self, *args, **kwargs):
        self.check_memory()
        return await super().handle(*args, **kwargs)
```

2. Update your Dockerfile CMD:
```dockerfile
CMD ["gunicorn", "main:app", "-k", "worker.MemoryLimitedWorker"]
```

This approach:
- Monitors memory usage per worker
- Automatically restarts workers that exceed memory limits
- Provides visibility into memory usage through logs

Both solutions are:
- Easy to implement
- Require minimal configuration
- Provide automatic recovery from memory issues
- Scale well with your application

Choose Solution 1 if you prefer a container-level approach with less code, or Solution 2 if you want more detailed control and monitoring of worker memory usage.

Remember to add `psutil` to your requirements.txt if using Solution 2:
```
psutil>=5.8.0
```

Let me know if you need any clarification on implementing either solution!

**Comment by YuriiMotov** [comment]
I think this question is more related to Gunicorn.

Let's close this for now to keep the discussions clean.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
